### DISCUSSION TITLE
Source filename in PMD 7

### DISCUSSION THREAD

**Discussion by justinstroud-campmeeker** [pmd/pmd#4865] [Q&A] [answered]
Hi all. New to PMD customization so pardon if this seems like a dumb question, but in PMD <7 you could get the source filename from the rule context. My context: I have XML files with a different extension and I have to bypass the rule if the source file being inspected does not have that extension. How do I grab the source filename from the rule in 7?

For the record, I'm working on a language module for the file types I'm working with but I need a quick fix for right now.

**Comment by jsotuyod** [comment, marked as answer]
It's not in the context, but you can get it from any node in the AST through `Node.getTextDocument().getFileId().getFileName()`.

**Reply by justinstroud-campmeeker** [reply]
That will work, although I am curious as to why it was removed from the RuleContext because that seemed to be a logical place for that bit of information. Aside from creating a new language module, is there a better (more centralized) way to exclude based on filename at a higher level than the visit(...) method? Oh...and thank you for your response!

**Reply by jsotuyod** [reply]
Well, the code language isn't really relative to the rule (rule context), but to the code itself (AST).

> Aside from creating a new language module, is there a better (more centralized) way to exclude based on filename at a higher level than the visit(...) method?

I'm honestly a little lost as to what you are doing… on the one hand, your original post said "I have XML files with a different extension", so I assumed your files are NOT `.xml` files but `.somethingelse`, but you are processing them as such either by having a custom language module register that extension; or by using the `--force-language` CLI option, but in either case, I'm unsure why you would also want to exclude some rules based on the very extension.

**Reply by justinstroud-campmeeker** [reply]
While I'm working on a better solution (I actually want to build a proper Salesforce metadata language module), I have a requirement to get Salesforce flow metadata validated. Those have an extension of .flow-meta.xml. I only want my rules to run against files with that 'extended' extension, not everything in the directory with a .xml extension. That's the abridged version.

I'm OK the fact that these files are more like documents than code, but some of them define processes that need to be checked against best practices.

**Reply by justinstroud-campmeeker** [reply]
To answer your last question though: They are technically XML files and the XML language module was my starting point in order to get some rules in place (if you're doing a data operation, require a fault branch, that sort of thing.) My ultimate goal is to build a language module for the Salesforce <whatever>-meta.xml files that we care about.

**Reply by jsotuyod** [reply]
Oh, I see. So technically they *are* `.xml` files, but also `flow-meta.xml`… yeah, we don't have anything better at the moment I fear. Even if you built a new language module, I'm unsure how PMD would decide which language to apply to the code (I'd guess it depends on which one is found first in the classpath…).

If you are up to it, this new language module could be a very interesting contribution. There is definitely interest in the community to have SF flow support, as per https://github.com/pmd/pmd/issues/3413, https://github.com/pmd/pmd/issues/4332, https://github.com/pmd/pmd/issues/3846 …

**Reply by justinstroud-campmeeker** [reply]
Yeah, I saw that there is definitely interest. I've heard that Salesforce has it "on the roadmap" but I'm not waiting around for it. I found a viable solution for 7 (thanks to you) and I have it working in 6.55 using the XML language module. My eventual goal (I'm repeating myself, sorry) is to build out another language module specific to Salesforce metadata, then build others (like flow, permissions, etc.) on top of that...similar to the way the XML processor works (how it has pom, xslt, etc. under it.) If I approach something close to what I'd be comfortable (as in, not hacky and cleared by my employer) sharing I will definitely be up for contributing.

**Reply by justinstroud-campmeeker** [reply]
I spoke with the powers that be, and I'm happy to help. I have some ideas in mind as to how the metadata module would be structured (I mentioned the XML module as a starting point.) There are some differences though that I want to incorporate, because while it is XML, the schema is a little weird and it could change at a moment's notice (although SF is pretty decent about keeping things relatively stable.) I'll check out the links you posted above.
